src/client/websocket/WebSocketManager.js

```diff
@@ -23,307 +23,77 @@ class WebSocketManager extends EventEmitter {
-      d: this.sequence,
-    }, true);
-
-    this.lastHeartbeatAck = false;
+  heartbeat() {
+    if (!this.connection) return this.debug('No connection to heartbeat');
+    return this.connection.heartbeat();
   }
 
   /**
-   * Sends a packet to the gateway
-   * @param {Object} data An object that can be JSON stringified
-   * @param {boolean} force Whether or not to send the packet immediately
+   * Emits a debug event
+   * @param {string} message Debug message
+   * @returns {void}
    */
-  send(data, force = false) {
-    if (force) {
-      this._send(data);
-      return;
-    }
-    this._queue.push(data);
-    this.doQueue();
-  }
-
-  destroy() {
-    if (this.ws) this.ws.close(1000);
-    this._queue = [];
-    this.status = Constants.Status.IDLE;
-  }
-
-  _send(data) {
-    if (this.ws.readyState === WebSocketConnection.WebSocket.OPEN) {
-      this.emit('send', data);
-      this.ws.send(data);
-    }
-  }
-
-  doQueue() {
-    const item = this._queue[0];
-    if (!(this.ws.readyState === WebSocketConnection.WebSocket.OPEN && item)) return;
-    if (this.remaining === 0) {
-      this.client.setTimeout(this.doQueue.bind(this), Date.now() + (this._remainingReset || 120e3));
-      return;
-    }
-    this._remaining--;
-    this._send(item);
-    this._queue.shift();
-    this.doQueue();
+  debug(message) {
+    return this.client.emit('debug', `[ws] ${message}`);
   }
 
   /**
-   * Run whenever the gateway connections opens up
+   * Destroy the client
+   * @returns {boolean} Whether or not destruction was successful
    */
-  eventOpen() {
-    this.client.emit('debug', 'Connection to gateway opened');
-    this.lastHeartbeatAck = true;
-    if (this.sessionID) this._sendResume();
-    else this._sendNewIdentify();
-  }
-
-  /**
-   * Sends a gateway resume packet, in cases of unexpected disconnections.
-   */
-  _sendResume() {
-    if (!this.sessionID) {
-      this._sendNewIdentify();
-      return;
-    }
-    this.client.emit('debug', 'Identifying as resumed session');
-    this.resumeStart = this.sequence;
-    const payload = {
-      token: this.client.token,
-      session_id: this.sessionID,
-      seq: this.sequence,
-    };
-
-    this.send({
-      op: Constants.OPCodes.RESUME,
-      d: payload,
-    });
-  }
-
-  /**
-   * Sends a new identification packet, in cases of new connections or failed reconnections.
-   */
-  _sendNewIdentify() {
-    this.reconnecting = false;
-    const payload = this.client.options.ws;
-    payload.token = this.client.token;
-    if (this.client.options.shardCount > 0) {
-      payload.shard = [Number(this.client.options.shardId), Number(this.client.options.shardCount)];
-    }
-    this.client.emit('debug', 'Identifying as new session');
-    this.send({
-      op: Constants.OPCodes.IDENTIFY,
-      d: payload,
-    });
-    this.sequence = -1;
+  destroy() {
+    if (!this.connection) return this.debug('Attempted to destroy WebSocket but no connection exists!');
```

Comment:
return `this.debug` returns `this.client.emit` which means you return based on if the `debug` event has listeners